File: 46-permutations/permutations.cpp
class Solution {
public:
    int n;
    vector<vector<int>>finalans;
    void find(vector<int>& nums,vector<int>&v){
        bool flagg=false;
        for(int i=0;i<n;i++){
            if(nums[i]==INT_MIN) continue;

            flagg=true;
            int orgnums=nums[i];

            v.push_back(nums[i]);
            nums[i]=INT_MIN;

            find(nums,v);

            nums[i]=orgnums;
            v.pop_back();
        }
        if(flagg==false) finalans.push_back(v);
    }
    vector<vector<int>> permute(vector<int>& nums) {
        n=nums.size();
        vector<int>v;
        find(nums,v);
        return finalans;
    }
};

File: 46-permutations/permutations.java
class Solution {
    int n;
    List<List<Integer>>ans=new ArrayList<>();
    void find(int[] visited,List<Integer>v,int[] nums){
        if(v.size()==nums.length){
            ans.add(new ArrayList<>(v));
            return;
        }
        for(int i=0;i<nums.length;i++){
            if(visited[i]==1) continue;
            v.add(nums[i]);
            visited[i]=1;
            find(visited,v,nums);
            v.remove(v.size()-1);
            visited[i]=0;
        }
    }
    public List<List<Integer>> permute(int[] nums) {
        n=nums.length;

        List<Integer>v=new ArrayList<>();
        int[] visited=new int[n];
        for(int i=0;i<n;i++){
            visited[i]=0;
        }
        find(visited,v,nums);
        return ans;
    }
}